src/backend/drm/mod.rs

```diff
@@ -1,38 +1,153 @@
+//! Surface creation requires a `crtc` (which cannot be the same as another existing `Surface`'s crtc),
+//! as well as a `Mode` and a set of `connectors`.
+//!
+//! smithay does not make sure that `connectors` are not already in use by another `Surface`. Overlapping `connector`-Sets may
+//! be an error or result in undefined rendering behavior depending on the `Surface` implementation.
+//!
+//! ## Surface
+//!
+//! A surface is a part of a `Device` that may output a picture to a number of connectors. It pumps pictures of buffers to outputs.
 //!
 //! On surface creation a matching encoder for your `encoder`-`connector` is automatically selected,
 //! if it exists, which means you still need to check your configuration.
 //!
-//! At last a [`Mode`](drm::control::Mode) needs to be selected,
-//! supported by the `crtc` in question.
+//! A surface consists of one `crtc` that is rendered to be the user. This is fixed for the `Surface`s lifetime and cannot be changed.
+//! A surface also always needs at least one connector to output the resulting image to as well as a `Mode` that is valid for the given connector.
+//!
+//! The state of a `Surface` is double-buffered, meaning all operations that chance the set of `connector`s or their `Mode` are stored and
+//! only applied on the next commit. `Surface`s do their best to validate these changes, if possible.
+//!
+//! A commit/page_flip may be triggered by any other method the `Surface` has, but needs to be documented as such.
+//! The most low-level `Surface`s also implement `RawSurface`, which let you trigger a commit explicitly.
+//!
+//! ## RawDevice
+//!
+//! A low-level device that produces not only `Surface`s, but `RawSurface`s.
+//!
+//! ## RawSurface
+//!
+//! RawSurfaces provide two additional functions related to rendering. `commit` and `page_flip`.
+//! Both attach the contents of a framebuffer, resulting in it's contents being displayed on the set connectors.
+//!
+//! `commit` also applies any pending changes to the current `Mode` and `connector`-Sets.
+//! You can easily check, if there are pending changes and avoid the costly commit by using `commit_pending`,
+//! which compares the current and pending state internally.
+//!
+//! # Rendering
+//!
+//! To perform actual rendering operations you need to render your image to a framebuffer.
+//! Several types of framebuffers exists.
+//!
+//! The simplest one - the `Dumbbuffer` is a simple bitmap, usually very slow and not hardware-accelerated.
+//! It should be avoided at any cost, but works on all devices and can be used directly with a `RawSurface`.
+//! Rendering to it can be done manually by coping bitmaps around (cpu rendering).
+//!
+//! Accelerated buffers are usually hardware manufacture dependent. A problem, which is mostly solved by
+//! the General Buffer Manager (gbm-)API. `gbm::Buffer`s can only be rendered to by egl. Appropriate hardware-accelerated
```

Comment:
> [...] can only be rendered to by egl

Is GBM incompatible with Vulkan, or is it just that Smithay does not have Vulkan support yet?

Reply:
Last time I looked there was no Vulkan extension in Mesa to create vulkan objects from gbm ones. Such an extension (like the one existing for egl) would be necessary to use Vulkan on top of gbm.

As it stands there is no universally supported extension, so I assume most users of gbm will only see egl as a viable option.

Reply:
Alright then.